aptos-move/framework/aptos-framework/sources/ResourceAccount.move

```diff
@@ -0,0 +1,130 @@
+/// ```
+/// public fun add_coin<X, Y>(lp: &LP, x: Coin<x>, y: Coin<y>) {
+///     if(!exists<LiquidityCoin<X, Y>(LP::Address(lp), LiquidityCoin<X, Y>)) {
+///         let mint, burn = Coin::initialize<LiquidityCoin<X, Y>>(...);
+///         move_to(&create_signer_with_capability(&lp.cap), LiquidityCoin<X, Y>{ mint, burn });
+///     }
+///     ...
+/// }
+/// ```
+module AptosFramework::ResourceAccount {
+    use Std::Errors;
+    use Std::Signer;
+    use Std::Vector;
+    use AptosFramework::Account;
+    use AptosFramework::SimpleMap::{Self, SimpleMap};
+
+    const ECONTAINER_NOT_PUBLISHED: u64 = 0;
+
+    struct Container has key {
+        store: SimpleMap<address, Account::SignerCapability>,
+    }
+
+    /// Creates a new resource account and rotates the authentication key to either
+    /// the optional auth key if it is non-empty (though auth keys are 32-bytes)
+    /// or the source accounts current auth key.
+    public(script) fun create_resource_account(
+        source: &signer,
+        seed: vector<u8>,
+        optional_auth_key: vector<u8>,
+    ) acquires Container {
+        let (signer, signer_cap) = Account::create_resource_account(source, seed);
+
+        let source_addr = Signer::address_of(source);
+        if (!exists<Container>(source_addr)) {
+            move_to(source, Container { store: SimpleMap::create() })
+        };
+
+        let container = borrow_global_mut<Container>(source_addr);
+        let signer_addr = Signer::address_of(&signer);
+        SimpleMap::add(&mut container.store, signer_addr, signer_cap);
+
+        let auth_key = if (Vector::is_empty(&optional_auth_key)) {
+            Account::get_authentication_key(source_addr)
+        } else {
+            optional_auth_key
+        };
+        Account::rotate_authentication_key_internal(&signer, auth_key);
+    }
+
+    /// When called by the resource account, it will retrieve the capability associated with that
+    /// account and rotate the account's auth key to 0x0 making the account inaccessible without
+    /// the SignerCapability.
+    public fun retrieve_resource_account_cap(
+        resource: &signer,
+        source_addr: address,
+    ): Account::SignerCapability acquires Container {
+        assert!(exists<Container>(source_addr), Errors::not_published(ECONTAINER_NOT_PUBLISHED));
+
+        let resource_addr = Signer::address_of(resource);
+        let (signer_cap, remove) = {
+            let container = borrow_global_mut<Container>(source_addr);
+            let (_resource_addr, signer_cap) = SimpleMap::remove(&mut container.store, &resource_addr);
+            (signer_cap, SimpleMap::length(&container.store) == 0)
+        };
+
+        if (remove) {
+            let container = move_from(source_addr);
+            let Container { store: store } = container;
+            SimpleMap::destroy_empty(store);
+        };
+
+        let empty_auth_key = Vector::empty();
+        while (Vector::length(&empty_auth_key) < 32) {
```

Comment:
Wow... Are there any better way of constructing the auth key for 0x0? If not, can you add comments to explain that this is constructing the auth key for 0?